pipeline doc - (#14414)
dotnet/docs: docs/standard/io/Buffers.md

# System.Buffers

- **Writing**
    - [IBufferWriter\<T\>](#ibufferwritert)
       - [Gotchas](#gotchas)
- **Reading**
    - [ReadOnlySequence\<T\>](#readonlysequencet)
       - [Accessing data](#accessing-data)
       - [Gotchas](#gotchas-1)
    - [SequenceReader\<T\>](#sequencereadert)
       - [Accessing data](#accessing-data-1)
       - [Gotchas](#gotchas-2)
   

> ## [IBufferWriter\<T\>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.ibufferwriter-1?view=netstandard-2.1)

Reviewer: All the API links should be changed to xrefs. For example: 

```suggestion
## <xref:System.Buffers.IBufferWriter%601>
```

> `IBufferWriter<T>` is a contract for synchronous buffered writing. At the lowest level the interface is extremely simple and allows you to get access to a `Memory<T>` or `Span<T>`, write to it and say how many `T` items were written.

Reviewer: ```suggestion
`IBufferWriter<T>` is a contract for synchronous buffered writing. At the lowest level, the interface is very simple and allows you to get access to a <xref:System.Buffers.MemoryManager%601.Memory> or <xref:System.Memory%601.Span>, write to it and say how many `T` items were written.
```

Reviewer: Why would I want to do that?

Author: @mairaw  simple is banned by the style guide, esp my style guide. See my commits if you want to see the type of changes I'm making.

> ```C#

Reviewer: ```suggestion
```csharp
```

void WriteHello(IBufferWriter<byte> writer)
{
    // Request at least 5 bytes
    Span<byte> span = writer.GetSpan(5);
    ReadOnlySpan<char> helloSpan = "Hello".AsSpan();
    int written = Encoding.ASCII.GetBytes(helloSpan, span);
    

>     // Tell the writer how many bytes we wrote

Reviewer: ```suggestion
    // Tell the writer how many bytes were written
```

    writer.Advance(written);
}
```

The preceding method requests a buffer of at least 5 bytes from the `IBufferWriter<byte>` using `GetSpan(5)` then writes bytes for the ASCII string "Hello" to the returned `Span<byte>`. It then calls `Advance(written)` to indicate how many bytes were written to the buffer. 

> This method of writing uses the `Memory<T>`/`Span<T>` buffer provided by the `IBufferWriter<T>`, but you can also use the [`Write`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.buffersextensions.write?view=netstandard-2.1) extension method to copy an existing buffer to the `IBufferWriter<T>`. `Write` does the work of calling `GetSpan`/`Advance` as appropriate, so there's no need to call `Advance` after writing.

Reviewer 2: We should have links to the API doc for GetSpan/GetMemory/Advance. Ex: https://docs.microsoft.com/en-us/dotnet/api/system.buffers.ibufferwriter-1.getmemory?view=netstandard-2.1#System_Buffers_IBufferWriter_1_GetMemory_System_Int32_

Reviewer 3: > PLEASE HOLD COMMENTS UNTIL IT'S READY FOR REVIEW

😆

Reviewer 2: Yeah, I stopped as soon as I saw that. Oops!

Author: NP. This is taking a back seat to the ASP.NET Core 3.0 release so I'll be slow doing updates. At some point I'll need one of you to get the bulk of the pipe code into a working project on a Git repo. That way I can migrate the code snippets to this doc and not have embedded snippets.

```C#
void WriteHello(IBufferWriter<byte> writer)
{
    byte[] helloBytes = Encoding.ASCII.GetBytes("Hello");
    
    // Write helloBytes to the writer, there's no need to call Advance here (Write does that)
    writer.Write(helloBytes);
}
```

[`ArrayBufferWriter<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraybufferwriter-1?view=netcore-3.0) is an implementation of `IBufferWriter<T>` whose backing store is a single contiguous array.

### Gotchas

- `GetSpan` and `GetMemory` return a buffer with at least the requested amount of memory. Don't assume exact buffer sizes.
- There is no guarantee that successive calls will return the same buffer or the same-sized buffer.
- You must request a new buffer after calling `Advance` to continue writing more data; you cannot write to a previously acquired buffer.

## [ReadOnlySequence\<T\>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1)

> ![image](https://user-images.githubusercontent.com/95136/64049773-cbd04600-cb2a-11e9-9d37-404488a2d6d5.png)

Reviewer: This needs alt text and be imported into our structure

`ReadOnlySequence<T>` is a struct that can represent a contiguous or discontiguous sequence of T. It can be constructed from:
1. A `T[]`
1. A `ReadOnlyMemory<T>`
1. A pair of linked list node [`ReadOnlySequenceSegment<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequencesegment-1?view=netstandard-2.1) and index to represent the start and end position of the sequence

> The 3rd representation is the most interesting one as it has performance implications on various operations on the `ReadOnlySequence<T>`:

Reviewer: ```suggestion
The third representation is the most interesting one as it has performance implications on various operations on the `ReadOnlySequence<T>`:
```

|`ReadOnlySequenceSegment<T>`|`GetPosition(long)`|`O(number of segments)`
|`ReadOnlySequenceSegment<T>`|`Slice(int, int)`|`O(number of segments)`
|`ReadOnlySequenceSegment<T>`|`Slice(SequencePostion, SequencePostion)`|`O(1)`

Because of this mixed representation, the `ReadOnlySequence<T>` exposes indexes as `SequencePosition` instead of an integer. A `SequencePosition` is an opaque value that represents an index into the `ReadOnlySequence<T>` where it originated. It consists of 2 parts, an integer and an object, what these 2 values represent are tied to the implementation of `ReadOnlySequence<T>`.

### Accessing data

The `ReadOnlySequence<T>` exposes data as an enumerable of `ReadOnlyMemory<T>`. Enumerating each of the segments can be done using a simple foreach:

```C#
long FindIndexOf(in ReadOnlySequence<byte> buffer, byte data)
{
    long position = 0;
    
    foreach (ReadOnlyMemory<byte> segment in buffer)
    {
        ReadOnlySpan<byte> span = segment.Span;
        var index = span.IndexOf(data);
        if (index != -1)
        {
            return position + index;
        }
        
        position += span.Length;
    }
    
    return -1;
}
```

> The above method searches each segment for a specific byte. If we need to keep track of each segment's `SequencePosition` then [`ReadOnlySequence.TryGet`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1.tryget?view=netcore-3.0) would be more appropriate. Let's change the above code to return a `SequencePosition` instead of an integer. This has the added benefit of allowing the caller to avoid a second scan to get the data at a specific index.

Reviewer 4: It took me a little while to work out what this mean:

> "allowing the caller to avoid a second scan"

There are two kinds of scans going on here: the `Span<byte>.IndexOf` calls, and then the scan through the segments in the sequence. My mental model was that the bulk of the scan-like work here is the `IndexOf` because that's going to touch every byte until it finds what it's looking for. (Also, so far in my work with `ReadOnlySequence<T>`, which I've been using in conjunction with pipelines, the majority have been single-segment, with multi-segment examples only appearing as you approach the end of a buffer. I've found my app gets better throughput with largish buffer sizes, ensuring that most of the time I'm in the single-segment case, so I naturally tend to think of that one as the hot path that's most perf-critical.)

So maybe qualify this as "...a second scan through the sequence's segments to get..."?

Processing a `ReadOnlySequence<T>` can be challenging since data may be split across multiple segments within the sequence. For the best performance, split code into a fast path, that deals with the single segment case, and a slow path that deals with the data split across segments. There are a few approaches that can be used to process data in multi-segmented sequences:

- Use the [`SequenceReader<T>`](#sequencereadert)
- Parse data segment by segment, keeping track of the `SequencePosition` and index within the segment parsed. This avoids unnecessary allocations but may be inefficient (especially for small buffers).
- Copy the `ReadOnlySequence<T>` to a contiguous array and treat it like a single buffer:
  - If the `ReadOnlySequence<T>` has a length less then 256, it may be reasonable to copy the data into a stack-allocated buffer (using the `stackalloc` keyword).
  - Copy the `ReadOnlySequence<T>` into a pooled array using `ArrayPool<T>.Shared`.
  - Use `ReadOnlySequence<T>.ToArray()`. This is not recommended in hot paths as it allocates a new `T[]` on the heap.

The following examples demonstrate some common cases for processing `ReadOnlySequence<byte>`:

> ##### Processing Binary Data

Reviewer: Is this really supposed to be an H5?

```suggestion
##### Processing binary data
```
